### PR TITLE
frieren: soft no-slip BC penalty (tau dot n_hat) for wall_shear_y/z gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Soft no-slip / no-penetration physics penalty on the wall-shear prediction.**

The wall-shear stress vector `tau` is, by physics, **tangent to the surface** at every point — it cannot have a component along the surface normal `n_hat`. The current SOTA recipe (PR #222) does not enforce this constraint at all. Kohaku's `--use-tangential-wallshear-loss` (PR #11) projects the prediction onto the tangent plane *before* computing MSE, which means any normal-component error is silently discarded — the model is never punished for predicting normal-direction garbage and never learns that the constraint exists. The latent `wallshear_pred_normal_rms` metric (already logged when tangential mode is on) shows the model does in fact violate this BC.

This experiment adds a **soft penalty term** instead of (or in addition to) a hard projection:

```
L = surface_loss + volume_loss + λ_normal · mean( (tau_pred · n_hat)^2 )
```

Computed in **physical (denormalized) space** for unit consistency with the existing tangential-loss code path. This teaches the network a hard physical prior with a single scalar hyperparameter and zero architectural change. Wall_shear_y/z account for ~70% of the abupt headline gap and are the most prone to normal-component leakage (because the dominant tau_x has clearer streamline alignment).

**Why this is orthogonal to current in-flight work:**
- #207 alphonse (AGC), #247 thorfinn (T_max=14), #262 nezuko (warmdown), #261 norman (Muon), #245 gilbert (EMA), #249 tanjiro (asinh), #244 emma (surf weight), #243 chihiro (aux rel-L2): all about **optimization or loss-magnitude scaling** — none enforce a physics constraint.
- #227 stark (tangent-frame prediction): predicts in 2D tangent basis instead of 3D — pod missing per Issue #248 (RBAC blocked). This soft-penalty variant requires no tangent frame and no pod plumbing.
- #224 fern (learned Fourier), #225 haku (mirror sym aug), #193 thorfinn (curvature sampling): input-side or augmentation work.

The penalty is a clean, single-line addition; if it works it stacks with all of the above.

## Instructions

**Code change to `target/train.py`:**

1. Add a `Config` flag `wallshear_normal_penalty: float = 0.0` and CLI flag `--wallshear-normal-penalty FLOAT` (default 0.0 = off).
2. In `train_step` after the existing `surface_loss` / `volume_loss` computation (lines ~1355–1362), add:

   ```python
   if wallshear_normal_penalty > 0.0:
       normals = batch.surface_x[..., 3:6]
       n_hat = F.normalize(normals.float(), dim=-1, eps=1e-8)
       ws_std = transform.surface_y_std[1:4]
       ws_mean = transform.surface_y_mean[1:4]
       ws_pred_phys = surface_pred_norm[..., 1:4].float() * ws_std + ws_mean
       normal_dot = (ws_pred_phys * n_hat).sum(dim=-1)
       mask_f = batch.surface_mask.float()
       penalty = (normal_dot.square() * mask_f).sum() / mask_f.sum().clamp_min(1.0)
       loss = loss + wallshear_normal_penalty * penalty
       metrics["wallshear_normal_penalty"] = float(penalty.detach().cpu().item())
   ```

3. Thread `wallshear_normal_penalty` through `train_step()` signature and the call site, exactly as the existing `wallshear_y_weight` / `wallshear_z_weight` (PR #66) flags are threaded.
4. Add the metric to the per-step logging dict (`metrics["wallshear_normal_penalty"]`) — this lets us watch the constraint be learned over training.

**Important interaction with existing `--use-tangential-wallshear-loss`:** Do **not** combine `--use-tangential-wallshear-loss` with `--wallshear-normal-penalty` in this experiment. The projection erases the normal component so the penalty would always be 0. Run with `--use-tangential-wallshear-loss false` (default).

## Sweep design (4 arms, 1 GPU each, 4 GPUs total)

Use `--wandb-group frieren-wallshear-normal-penalty`.

Single-process per GPU with `--seed=-1`. Same SOTA recipe as PR #222 base config in **all** arms (4L/512d, Lion, lr=1e-4, lr_warmup_epochs=1, batch=4, ema=0.999):

| Arm | GPU | wallshear_normal_penalty (λ) | Run name suffix |
|---:|---:|---:|---|
| A (control) | 0 | 0.0 | `arm-A-ctrl` |
| B (light) | 1 | 0.1 | `arm-B-l0.1` |
| C (medium) | 2 | 0.5 | `arm-C-l0.5` |
| D (strong) | 3 | 2.0 | `arm-D-l2.0` |

The control arm is critical — same seed handling, same 4L/512d/Lion stack, no penalty — to control for any seed/HW noise. Compare arms B/C/D against arm A and against the published bar.

### Reproduce command (single arm; replicate per GPU with different `--wallshear-normal-penalty`)

```bash
cd target/
CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent frieren \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --wallshear-normal-penalty 0.1 \
  --seed -1 \
  --wandb-group frieren-wallshear-normal-penalty \
  --kill-thresholds "3000:train/loss<5,3000:train/grad/global_norm<300"
```

Run the 4 arms in parallel, one per GPU. **9 epochs** matches PR #222.

## Acceptance criteria

- **Merge bar:** any arm beating `val_primary/abupt_axis_mean_rel_l2_pct = 9.291%` at best epoch is mergeable.
- **Health gate:** `wallshear_normal_penalty` metric should monotonically decrease — if it stays flat or increases, the penalty is being ignored and we have a bug.
- **Stability gate:** if any arm trips the kill threshold (grad>300 at step 3000), that's a useful signal — note which λ failed and report the trajectory before the kill.

## Baseline (current best on `yi` — must beat 9.291%)

| Metric | yi best | PR | W&B |
|---|---:|---|---|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **9.2910** | #222 (fern, lr_warmup_epochs=1) | `ut1qmc3i` |
| `val_primary/surface_pressure_rel_l2_pct` | 5.8707 | #222 | `ut1qmc3i` |
| `val_primary/wall_shear_rel_l2_pct` | 10.3423 | #222 | `ut1qmc3i` |
| `val_primary/volume_pressure_rel_l2_pct` | 5.8789 | #222 | `ut1qmc3i` |

AB-UPT reference: surface_pressure=3.82, wall_shear=7.29 (this is the gap we're attacking), volume_pressure=6.08 (already beaten).

## Reporting

Reply with the standard results table on this PR when all 4 arms reach 9 epochs (or are killed):

| Arm | λ | best epoch | val_abupt | surf_pres | vol_pres | wall_shear | wallshear_normal_penalty (final) | run id |

Plus a short interpretation of how the penalty trajectory tracks against the abupt metric.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #279 because ADVISOR: Closing this PR before launch. I missed an existing experiment when designing the hypothesis — apologies, frieren.

**PR #201 (nezuko, 2026-05-01) already tested this exact penalty form** on the surface side and closed it with a hard-data finding:

> GT labels have RMS(tau dot n) / |tau| = 12% (mean=0.113, RMS=0.364, p99=1.722). DrivAerML mesh-discretized normals are not exactly orthogonal to the analytic surface, so the continuum no-slip constraint (tau dot n = 0) directly contradicts the ground truth. There is no lambda where this penalty can help:
>   - Small lambda: NaN cascade (>200 grad skips)
>   - Large lambda: forces tau_pred dot n -> 0, but labels have tau dot n ~= 0.36 -> 2x train loss inflation, 4x val regression
> Verdict: NEGATIVE.

The "tau is tangent to the surface" prior is geometrically true but **not satisfied by the discretized normals shipped with the dataset** — running this penalty would deterministically make the model worse, exactly as nezuko's λ=0.1 arm showed (val_abupt 63.09 vs control 15.55).

Reassigning you to a clean hypothesis now. No fault on you — this was an advisor mistake; should have grepped EXPERIMENTS_LOG.md for "no-slip" before writing the PR. Closing..
